src/main/java/org/xbill/DNS/DNSSession.java

```diff
@@ -0,0 +1,129 @@
+package org.xbill.DNS;
+
+import static java.lang.String.format;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import org.xbill.DNS.exceptions.AdditionalDetail;
+import org.xbill.DNS.exceptions.InvalidZoneDataException;
+import org.xbill.DNS.exceptions.LookupFailedException;
+import org.xbill.DNS.exceptions.RedirectOverflowException;
+import org.xbill.DNS.exceptions.ServerFailedException;
+
+/**
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
+ * and it's behaviour can be modified by calling the setter methods.
+ */
+public class DNSSession {
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
+
+  private final Resolver resolver;
+  private volatile int maxRedirects = DEFAULT_MAX_ITERATIONS;
```

Comment:
Why volatile?

Reply:
It will ensure that reads after a write will get the new value even if it is in a different thread. So, if thread A calls `setMaxRedirects()` with a new value and then thread B calls `lookupAsync()` the updated value will be visible to thread B. Without volatile, thread B might hold a cached copy of the old value indefinitely and not ever see the update.

Reply:
Since we have moved to a builder pattern now, this point is moot.